src/serde_impl.rs

```diff
@@ -0,0 +1,318 @@
+use ff::PrimeField;
+use group::GroupEncoding;
+use serde_crate::{
+    de::Error as DeserializeError, Deserialize, Deserializer, Serialize, Serializer,
+};
+
+use crate::{
+    curves::{EpAffine, EqAffine},
+    fields::{Fp, Fq},
+};
+
+const ERR_CODE: &str = "deserialized bytes don't encode a field element";
+
+/// Serializes bytes to human readable or compact representation.
+///
+/// Depending on whether the serializer is a human readable one or not, the bytes are either
+/// encoded as a hex string or a list of bytes.
+fn serialize_bytes<S: Serializer>(bytes: [u8; 32], s: S) -> Result<S::Ok, S::Error> {
+    if s.is_human_readable() {
+        hex::serde::serialize(bytes, s)
+    } else {
+        bytes.serialize(s)
+    }
+}
+
+/// Deserialize bytes from human readable or compact representation.
+///
+/// Depending on whether the deserializer is a human readable one or not, the bytes are either
+/// decoded from a hex string or a list of bytes.
+fn deserialize_bytes<'de, D: Deserializer<'de>>(d: D) -> Result<[u8; 32], D::Error> {
+    if d.is_human_readable() {
+        hex::serde::deserialize(d)
+    } else {
+        <[u8; 32]>::deserialize(d)
+    }
+}
+
+impl Serialize for Fp {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        serialize_bytes(self.to_repr(), s)
+    }
+}
+
+impl<'de> Deserialize<'de> for Fp {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let bytes = deserialize_bytes(d)?;
+        match Fp::from_repr(bytes).into() {
+            Some(fq) => Ok(fq),
+            None => Err(D::Error::custom(ERR_CODE)),
+        }
+    }
+}
+
+impl Serialize for Fq {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        serialize_bytes(self.to_repr(), s)
+    }
+}
+
+impl<'de> Deserialize<'de> for Fq {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let bytes = deserialize_bytes(d)?;
+        match Fq::from_repr(bytes).into() {
+            Some(fq) => Ok(fq),
+            None => Err(D::Error::custom(ERR_CODE)),
+        }
+    }
+}
+
+impl Serialize for EpAffine {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        serialize_bytes(self.to_bytes(), s)
+    }
+}
+
+impl<'de> Deserialize<'de> for EpAffine {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let bytes = deserialize_bytes(d)?;
+        match EpAffine::from_bytes_unchecked(&bytes).into() {
```

Comment:
The changes addressing @daira's earlier comments were lost during the rebase:
```suggestion
        match EpAffine::from_bytes(&bytes).into() {
```